### DISCUSSION TITLE
Allow throw statements on the right hand of the null coalescening assignment operator

### DISCUSSION THREAD

**Discussion by john-h-k** [dotnet/csharplang#2091] [Language Ideas] [answered]
In C# 8.0, this is valid code:
```
string x = null;
x = x ?? throw new Exception();"
```
However, for some reason, this code is not
```
string x = null;
x ??= throw new Exception
```
which would be a single line null check, but it generates CS8115 "A throw expression is not allowed in this context".

## Suggestion:
Either
* Allow throw expressions on the right side of the null coalescent assignment operator, semantically equivalent to `variable = variable ?? throw expression`
* Add a new, similar operator or use the null coalescing operator that an expression `variable -- operator -- flow change` is a valid statement, e.g replacing the `x ??= throw ...` line above with `x ?? throw...`

**Comment by yaakov-h** [comment]
```csharp
_ = x ?? throw new Exception();
```

is already valid, if you want a single line.

**Comment by DavidArno** [comment]
Whilst I can see a reason for making the compiler support `x ??= throw new Exception` for consistency, to my mind it is an incredibly unclear piece of "clever programming". Even if it were supported, I'd still prefer to write it as:
```cs
if (x is null) throw new Exception();
```
as the intent of the code is then clear. I think this a case of "just because you can [or in theory could] express code in a certain way doesn't mean you should".

**Comment by 333fred** [comment, marked as answer]
I won't say we can't be convinced to change our minds here, but this was a very explicit decision: https://github.com/dotnet/csharplang/blob/c81b99f4c965dbb6e9819a10a2e010db4c859f6f/meetings/2018/LDM-2018-07-16.md#throw-expression-on-the-right-hand-side. Our reasoning basically agreed with @DavidArno's above, we didn't see the benefits over a simple if statement. You're not trying to assign anything at all, you're doing a null check and throwing.

**Comment by gafter** [comment]
The code
``` c#
x ??= y;
```
is **not** equivalent to
``` c#
x = x ?? y;
```
rather it is equivalent to
``` c#
x ?? x = y;
```
Which obviously would not permit the throw expression for `y`.

**Reply by bernd5** [reply]
good explanation 👍
